### DISCUSSION TITLE
Elasticsearch ETL process erroring on successful 201 response

### DISCUSSION THREAD

**Discussion by nickchampion** [ravendb/ravendb#15343] [General]
Hi,

I have just got the Elasticsearch ETL working and I think the integration is erroring when it should not be, my index is being updated as expected so the API calls to ES are succeeding, this could be the Elasticsearch client (https://github.com/elastic/elasticsearch-net/blob/main/src/Elastic.Clients.Elasticsearch/Api/BulkResponse.cs)

Seems to me a 201 response should be treated as a successful operation, are there any implications of these errors on the RavenDB side because the error in the studio says the current batch has been stopped 

`Load error ratio is too high (errors: 140, successes: 0). Could not tolerate load error ratio and stopped current ETL batch`

Thanks

```
Raven.Server.Exceptions.ETL.ElasticSearch.ElasticSearchLoadException: Failed to index data to 'hctr_inv_records_1_0_1' index. Debug information: Invalid NEST response built from a successful (200) low level call on POST: /hctr_inv_records_1_0_1/_bulk?refresh=wait_for
# Invalid Bulk items:
  operation[0]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: YsvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[1]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: Y8vlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[2]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: ZMvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[3]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: ZcvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[4]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: ZsvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[5]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: Z8vlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[6]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: aMvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[7]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: acvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[8]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: asvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[9]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: a8vlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[10]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: bMvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[11]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: bcvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[12]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: bsvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[13]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: b8vlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[14]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: cMvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[15]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: ccvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[16]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: csvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[17]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: c8vlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[18]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: dMvlvIQB_TIgrqbDbrXN _version: 1 error: 
  operation[19]: index returned 201 _index: hctr_inv_records_1_0_1 _type:  _id: dcvlvIQB_TIgrqbDbrXN _version: 1 error: 
# Audit trail of this API call:
 - [1] HealthyResponse: Node: https://hctr-elastic.es.eu-west-2.aws.cloud.es.io/ Took: 00:00:00.9772534
# Request:
<Request stream not captured or already read to completion by serializer. Set DisableDirectStreaming() on ConnectionSettings to force it to be set on the response.>
# Response:
<Response stream not captured or already read to completion by serializer. Set DisableDirectStreaming() on ConnectionSettings to force it to be set on the response.>

   at Raven.Server.Documents.ETL.Providers.ElasticSearch.ElasticSearchEtl.ThrowElasticSearchLoadException(String message, ServerError serverError, Exception originalException, String debugInformation) in C:\Builds\RavenDB-Stable-5.4\54012\src\Raven.Server\Documents\ETL\Providers\Elasticsearch\ElasticSearchEtl.cs:line 263
   at Raven.Server.Documents.ETL.Providers.ElasticSearch.ElasticSearchEtl.LoadInternal(IEnumerable`1 records, DocumentsOperationContext context, EtlStatsScope scope)
   at Raven.Server.Documents.ETL.EtlProcess`6.Load(IEnumerable`1 items, DocumentsOperationContext context, TStatsScope stats) in C:\Builds\RavenDB-Stable-5.4\54012\src\Raven.Server\Documents\ETL\EtlProcess.cs:line 461
```

**Comment by arekpalinski** [comment]
Hi Nick,

We're erroring because according to the Elasticsearch API the response was not valid:

https://github.com/ravendb/ravendb/blob/6739f61cb5b6c68603b2d9d0c8b398ec1fd59843/src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticSearchEtl.cs#L141-L143

Although I don't see the root cause of that based on the message you sent. There was not `serverError` and also don't see what was `originalException`

https://github.com/ravendb/ravendb/blob/6739f61cb5b6c68603b2d9d0c8b398ec1fd59843/src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticSearchEtl.cs#L257-L265

Do you have a way to send us the repro so we could try it locally?

**Comment by nickchampion** [comment]
hi @arekpalinski if I were to set up a database for you on our cluster with the relevant collections and ETL configuration would that help?

**Reply by nickchampion** [reply]
Ok I have set up a database with the ETL configuration, could you send me an email address so I can send you the certificate and endpoint please @arekpalinski

**Reply by ayende** [reply]
support@ravendb.net

**Comment by nickchampion** [comment]
HI @arekpalinski I have set up the database for you with the ETL process enabled and failing to our ElasticCloud cluster, you can view the ETL setup here

https://a.hctr-dev.hectare.ravendb.cloud/studio/index.html#databases/tasks/ongoingTasks?&database=ETLRepro

Certificate to access this DB was sent to support@ravendb.net

Let me know if you need anything else from me.

Thanks

**Comment by arekpalinski** [comment]
The root cause was https://github.com/elastic/elasticsearch-net/issues/6154. It's about using latest v7.17.x .NET client when talking to v8.x Elasticsearch server (currently there isn't v8 .NET client generally available yet).

In order to make that work there is a need to enable compatibility mode: 
https://www.elastic.co/guide/en/elasticsearch/client/net-api/7.17/connecting-to-elasticsearch-v8.html#enabling-compatibility-mode

We did it via setting env variable `ELASTIC_CLIENT_APIVERSIONING: true`. In the next RavenDB release it will be available as option when defining a connection string to Elasticsearch: 
https://issues.hibernatingrhinos.com/issue/RavenDB-19599/Elasticsearch-ETL-Allow-to-enable-compatibility-mode
